### PR TITLE
Facematch

### PR DESCRIPTION
**Pull request by RohanDas-max**
1. face-match api done

### CODE REVIEW

File: pkg/handler/faceMatch.go [hunk 1]
```diff
@@ -0,0 +1,29 @@
...
+
+func FaceMatch(c *gin.Context) {
+	var r resources.FaceMatchReqBody
+	if err := c.ShouldBindJSON(&r); err != nil {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{
+			"errorMessage": err.Error(),
+		})
+		return
+	}
+	score, err := services.FaceMatchService(c, &r)
```
Comment (iamargus95) [inline pkg/handler/faceMatch.go:19]: new line

File: pkg/handler/faceMatch.go [hunk 2]
```diff
@@ -0,0 +1,29 @@
...
+	if err != nil {
+		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{
+			"errorMessage": err.Error(),
+		})
+		return
+	}
+	c.JSON(http.StatusOK, gin.H{
```
Comment (iamargus95) [inline pkg/handler/faceMatch.go:26]: new line

File: pkg/handler/faceMatch_test.go [hunk 3]
```diff
@@ -0,0 +1,59 @@
...
+	"testing"
+
+	"github.com/gin-gonic/gin"
+	"github.com/joho/godotenv"
+	"github.com/rohandas-max/ekyc/pkg/caching"
+	"github.com/rohandas-max/ekyc/pkg/database"
+	"github.com/rohandas-max/ekyc/pkg/middleware"
+)
+
+func TestFaceMatch(t *testing.T) {
```
Comment (iamargus95) [inline pkg/handler/faceMatch_test.go:20]: use t.Setenv in all tests.

File: pkg/handler/faceMatch_test.go [hunk 4]
```diff
@@ -0,0 +1,59 @@
...
+	tests := []struct {
+		testName string
+		BodyData []byte
+		want     int
+	}{
+		{
+			testName: "200 for correct input",
+			BodyData: []byte(`{"Image1":"25480ab1-5520-4f38-badf-b42e8fb694f0","Image2:"29b5afff-5937-43ba-983c-f2b602a44cf3"}`),
+			want:     200,
+		}, {
```
Comment (iamargus95) [inline pkg/handler/faceMatch_test.go:39]: rectify indentation here.
and also add tests to fulfill empty cases and same uuids in both fields.

File: pkg/handler/faceMatch_test.go [hunk 5]
```diff
@@ -0,0 +1,59 @@
...
+			want:     400,
+		},
+	}
+
+	for _, tc := range tests {
+		t.Run(tc.testName, func(t *testing.T) {
+			rec := httptest.NewRecorder()
+			c, _ := gin.CreateTestContext(rec)
+			requestData := tc.BodyData
+			c.Request, _ = http.NewRequest(http.MethodPost, "api/v1/signup", bytes.NewBuffer(requestData))
```
Comment (iamargus95) [inline pkg/handler/faceMatch_test.go:51]: why is it hitting api/v1/signup?

File: pkg/handler/uploadImage.go [hunk 6]
```diff
@@ -10,15 +10,22 @@ import (
...
+
 	file, fileHeader, err := c.Request.FormFile("Image")
 	if err != nil {
 		c.AbortWithStatusJSON(http.StatusBadRequest, gin.H{
 			"errorMessage": err.Error(),
 		})
 		return
 	}
-	uuid, err := services.UploadService(c, typ, file, *fileHeader)
+	uuid, err := services.UploadService(c, r.Type, file, *fileHeader)
```
Comment (iamargus95) [inline pkg/handler/uploadImage.go:28]: new line

File: pkg/models/fileMetadata.go [hunk 7]
```diff
@@ -1,10 +1,10 @@
 package models
 
 type FileMetaData struct {
```
Comment (iamargus95) [inline pkg/models/fileMetadata.go:3]: why no gorm.model?

File: pkg/routers/router.go [hunk 8]
```diff
@@ -23,8 +23,8 @@ func Router() *gin.Engine {
 	v1.Use(middleware.IsAuthorized)
 	{
 		v1.POST("/image", handler.UploadImage)
+		v1.POST("/face-match", handler.FaceMatch)
 	}
-
 	r.NoRoute(func(ctx *gin.Context) {
```
Comment (iamargus95) [inline pkg/routers/router.go:28]: Move to a separate file under controllers.

File: pkg/services/faceMatchService.go [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+
+func FaceMatchService(c *gin.Context, r *resources.FaceMatchReqBody) (string, error) {
+	score := rand.Intn(100)
+	clientId, _, _, _ = middleware.UserDetail.Get()
+	if isUUIDExists(*r, clientId) {
+		cachedScore, err := cachedScore(c, "score", score)
+		if err != nil {
+			return "", err
+		}
+		if cachedScore == "" {
```
Comment (iamargus95) [inline pkg/services/faceMatchService.go:25]: new line

File: pkg/services/faceMatchService.go [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+			return strconv.Itoa(score), nil
+		}
+		return cachedScore, nil
+	}
+	return "", errors.New("invalid uuid")
+}
+
+func cachedScore(c *gin.Context, key string, value int) (string, error) {
+	res, err := caching.Get(c, key)
+	if err != nil && res == "" {
```
Comment (iamargus95) [inline pkg/services/faceMatchService.go:35]: why check for res == ""?

File: pkg/services/faceMatchService.go [hunk 11]
```diff
@@ -0,0 +1,50 @@
...
+		err := caching.Set(c, key, value)
+		if err != nil {
+			return "", err
+		}
+	}
+	return res, nil
+}
+
+func isUUIDExists(r resources.FaceMatchReqBody, clientId int) bool {
```
Comment (iamargus95) [inline pkg/services/faceMatchService.go:44]: move to helper functions.

File: pkg/services/faceMatchService.go [hunk 12]
```diff
@@ -0,0 +1,50 @@
...
+		}
+		if cachedScore == "" {
+			return strconv.Itoa(score), nil
+		}
+		return cachedScore, nil
+	}
+	return "", errors.New("invalid uuid")
+}
+
+func cachedScore(c *gin.Context, key string, value int) (string, error) {
```
Comment (iamargus95) [inline pkg/services/faceMatchService.go:33]: move to helper func or utils.

File: pkg/services/faceMatchService_test.go [hunk 13]
```diff
@@ -0,0 +1,140 @@
...
+	"github.com/rohandas-max/ekyc/pkg/resources"
+	"github.com/stretchr/testify/assert"
+)
+
+func TestFaceMatchService(t *testing.T) {
+	if err := godotenv.Load("../../.env"); err != nil {
+		log.Fatal(err)
+	}
+	dsn := fmt.Sprintf("host=%v user=%v password=%v dbname=%v port=%v sslmode=%v TimeZone=%v",
+		os.Getenv("DB_HOST"), os.Getenv("DB_USER"), os.Getenv("DB_PASSWORD"), os.Getenv("DB_DBNAME"), os.Getenv("DB_PORT"), os.Getenv("DB_SSLMODE"), os.Getenv("DB_TIMEZONE"))
```
Comment (iamargus95) [inline pkg/services/faceMatchService_test.go:26]: t.Setenv.

File: pkg/services/faceMatchService_test.go [hunk 14]
```diff
@@ -0,0 +1,140 @@
...
+	database.SetUp(dsn)
+	token := os.Getenv("TOKEN")
+	db, _ := strconv.Atoi(os.Getenv("REDIS_DB"))
+	caching.SetUp(os.Getenv("REDIS_ADDR"), os.Getenv("REDIS_PASS"), db)
+	tests := []struct {
```
Comment (iamargus95) [inline pkg/services/faceMatchService_test.go:31]: Improve tests by adding more cases. Have a case for same uuid's

File: pkg/services/faceMatchService_test.go [hunk 15]
```diff
@@ -0,0 +1,140 @@
...
+			c.Request = httptest.NewRequest(http.MethodPost, "/api/v1/face-match", nil)
+			c.Request.Header.Set("Content-Type", "appliaction/json")
+			c.Request.Header.Add("Authorization", "Bearer "+token)
+			middleware.IsAuthorized(c)
+			gotRes, gotErr := FaceMatchService(c, tc.r)
+			defer rec.Result().Body.Close()
+			if gotErr.Error() != tc.wantErr {
+				t.Errorf("FaceMatchService()=%v,%v,want = %v", gotRes, gotErr, tc.wantErr)
+			}
+		})
```
Comment (iamargus95) [inline pkg/services/faceMatchService_test.go:67]: No checking for gotRes?

File: pkg/services/faceMatchService_test.go [hunk 16]
```diff
@@ -0,0 +1,140 @@
...
+	}
+
+}
+
+func TestIsUUIDExists(t *testing.T) {
+	if err := godotenv.Load("../../.env"); err != nil {
+		log.Fatal(err)
+	}
+	dsn := fmt.Sprintf("host=%v user=%v password=%v dbname=%v port=%v sslmode=%v TimeZone=%v",
+		os.Getenv("DB_HOST"), os.Getenv("DB_USER"), os.Getenv("DB_PASSWORD"), "ekyc", os.Getenv("DB_PORT"), os.Getenv("DB_SSLMODE"), os.Getenv("DB_TIMEZONE"))
```
Comment (iamargus95) [inline pkg/services/faceMatchService_test.go:107]: t.Setenv

File: pkg/services/imageUploadService.go [hunk 17]
```diff
@@ -16,27 +16,28 @@ func UploadService(c *gin.Context, fileType string, file multipart.File, fileHea
...
+func upload(c *gin.Context, typ string, file multipart.File, fileHeader *multipart.FileHeader, id int, bucketName string) (string, error) {
 	info, err := objectStorage.UploadImageToBucket(c, fileHeader.Filename, bucketName, file, fileHeader.Size)
 	if err != nil {
-		return minio.UploadInfo{}, err
+		return "", err
 	}
-	database.InsertIntoFileMetadata(typ, filepath.Ext(info.Key), int(info.Size/1024), id)
-	return info, nil
+	uuid := uuid.New().String()
+	database.InsertIntoFileMetadata(typ, filepath.Ext(info.Key), uuid, int(info.Size/1024), id)
```
Comment (iamargus95) [inline pkg/services/imageUploadService.go:41]: db transaction error/

**Review by iamargus95 — CHANGES_REQUESTED** [review]

**Review by iamargus95 — APPROVED** [review]

**Comment by kalpeshdeo** [comment]
Please address the comments above
